docker/setup.sh

```diff
@@ -0,0 +1,21 @@
+#!/bin/bash
```

Comment:
Je ne suis pas d'accord par principe avec l'automatisation des run de migration, même si on est en théorie ici sur un build propre.
Je préfère que tu documentes les trois commandes a lancer que de lancer un script automagique :)

Reply:
En fait ce script est éxécuté *dans* le container, ce n'est pas à l'utilisateur de le lancer.
https://github.com/simde-utc/portail/blob/a12ed0571dd9435f88e90ac1e57a3fedc1506735/docker/app.Dockerfile#L64
Il sert à installer la BDD lors du premier lancement parce qu'il faut que les 2 containers soient lancés en même temps pour communiquer (impossible de le faire au build)

Reply:
Tu peux pas build tes deux containers avant , et lancer ces commandes ensuite ? 
un petit `docker-compose run -rm php artisant blabla ?`

Reply:
C'est un peu bête puisque dans le container ne sera pas utilisable tant que ces commandes n'auront pas été tapées :frowning_face: 

On peut effectivement faire comme ça, n'importe quand quand le container est lancé :
```console
$ docker-compose exec portal_app_1 php artisant blabla
```
avec `portal_app_1` le nom du container

Reply:
En l'occurence c'est le nom du service qui serait a utiliser et non le nom de l'instance, donc ` docker-compose exec app php artisant blabla` ;) 
Mais oui, quand tu déploie il y a quelques étapes a passer. Que ce soit injecter ses credentials quelque part, changer une variable d'env ailleurs, ou jouer les migrations / seeders etc c'est a la charge du gars/script qui va déployer. Le docker n'est pas censé avoir cette autonomie.

Reply:
> En l'occurence c'est le nom du service qui serait a utiliser et non le nom de l'instance, donc ` docker-compose exec app php artisant blabla` ;)

Oups, j'ai confondu avec `docker exec` :sweat_smile: 

Pour Docker, une image représentant une version d'un service, il m'est paru normal d'effectuer les migrations lors de la création d'un nouveau container. Puisque ces containers sont immuables, ils sont aussi associés à un état de schéma base de données, et la mise à jour de ce schéma devrait donc se faire avec la mise à jour du container. Étant parti sur une version de production, cela m'a semblé le plus judicieux (ça suit l'« Infrastructure as Code »).

Après, comme toi et @NastuzziSamy l'avez assez bien montré, vous utilisez surtout Docker pour le développement, ce qui explique sûrement nos divergences :smile: Je comprends ta vision des choses, mais encore une fois c'est une version de production, et est donc destiné à fonctionner clé en main, avec le moins d'intervention de la part de l'opérateur, afin d'éviter de créer des problèmes dans l'infrastructure. La version développeur sera bien plus pratique, sans pour autant balancer tout dans le dossier du projet et laisser le dev trouver les bonnes commandes pour ne rien casser (je dis ça parce qu'en tant que nouvel utilisateur du projet, j'ai eu un peu de mal à comprendre fonctionne l'ensemble, avec ou sans Docker)... :wink:

Reply:
En fait non ce que tu décris ne suit pas l'IaC, cette dernière ne défini pas si le modèle de donnée doit être a la même version que le code métier associé (tant qu'il reste fonctionnel, évidemment). Tu veux absolument utiliser les mots "de production" alors qu'il n'y a aucune production de ce docker ^^ Soit ! Je veux m'en servir comme exemple : Tu as une infinité de cas ou les migrations ne sont pas forcément à jouer en synchronisation avec les déploiements de code, c'est meme une bien meilleure pratique de dissocier les deux actions pour éviter les problèmes :)
C'est notamment pour ca que je te demande (et t'indique en conseil) de rendre manuelles ces actions pour laisser le moment de leur execution à la discretion de l'utilisateur final.
( lancer un `docker-compose exec app composer install && php artisan migrate:refresh --seed --force && php artisan portail:clear` c'est pas la mer à boire :P )

Reply:
J'ai appliqué les migrations de cette façon pour pouvoir lancer toute l'infrastructure en une commande :muscle: Blague à part, j'ai juste cherché à intégrer le `init.sh` ainsi que les instructions du README dans la configuration Docker, puisque je ne connais pas trop les technos de ce projet (je suis plus JS back/front :sweat_smile:)

J'avoue ne pas avoir cherché à comprendre, je pensais qu'en effectuant les migrations ça appliquerait tout seul les bonnes (en excluant celles à ne pas déployer).

Merci pour l'info, c'est noté :wink:

> ( lancer un `docker-compose exec app composer install && php artisan migrate:refresh --seed --force && php artisan portail:clear` c'est pas la mer à boire :P )

T'as essayé avec Heroku :see_no_evil:  ?